crates/adapters/src/controller/mod.rs

```diff
@@ -1146,6 +1173,28 @@ impl ControllerInner {
         }
     }
 
+    /// Waits for input `step` to settle.
+    ///
+    /// An input step must settle before we can commit any of the output.
+    /// Otherwise, if there is a crash, we might have output for which we don't
+    /// know the corresponding input, which would break fault tolerance.
+    fn wait_to_settle(self: &Arc<Self>, step: Step) {
+        if !self.status.step_is_settled(step) {
+            let mut guard = self.unsettled_step.lock().unwrap();
+            loop {
+                let unsettled_step = self.status.unsettled_step().unwrap();
+                if unsettled_step > *guard {
+                    *guard = unsettled_step;
+                    self.step_settled.notify_all();
+                }
+                if unsettled_step < step {
```

Comment:
It's confusing because "unsettled_step" reverses the sense of what we care about. We want to wait for `step` to settle, which means that the first step that is *not* settled (unsettled) has to be beyond `step`, hence the condition.

The reason that we track the first unsettled step, rather than the last settled step, is that there might not be any settled steps at all, if we just initialized. We are left with at least the following choices, other than this one:
* Use `Option<Step>` to track the first settled step. It can be `None` before any steps settle. This is an annoying representation because of the special cases.
* Make the first step number 1 rather than number 0, and use 0 for the first settled step before any settle. This is viable and I'd be willing to do it this way.
* Use a signed type, e.g. `i64`, for `Step`, and use -1 for the first settled step before any settle. This is viable and I'd be willing to do it this way.

Basically, I'm not sure of the best representation for step numbers, and I'm happy to adopt a different representation at this point.